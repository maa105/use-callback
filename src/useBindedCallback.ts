import { useRef } from 'react'
import { Handler, UseBindedCallbackState } from './typings'

const useBindedCallback = (handler: Handler, ...bindArgs: any[]) => {
  const ref = useRef<UseBindedCallbackState>()
  const current = ref.current
  if (current) {
    current.handler = handler
    current.bindArgs = bindArgs
    return current.callback
  }
  const mutable: any = {
    handler,
    bindArgs
  }
  const callback = function (this: any, ...args: any[]) {
    return mutable.handler.call(this, ...mutable.bindArgs, ...args)
  }
  mutable.callback = callback
  ref.current = mutable
  return callback
}

export default useBindedCallback
