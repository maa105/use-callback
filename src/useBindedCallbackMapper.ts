import { useRef } from 'react'
import {
  Handler,
  KeyGetter,
  MapFunction,
  Mapper,
  UseBindedCallbackMapperState,
  UseCallbackMapperItemsState,
  UseCallbackMapperItemState
} from './typings'

const useBindedCallbackMapper = (
  handler: Handler,
  keyGetter: string | KeyGetter,
  ...bindArgs: any[]
): Mapper => {
  const ref = useRef<UseBindedCallbackMapperState>()
  const current = ref.current
  if (current) {
    current.handler = handler
    current.bindArgs = bindArgs
    return current.mapper
  }
  const itemsMutable: UseCallbackMapperItemsState = {}
  const getKey: KeyGetter =
    typeof keyGetter === 'string' ? (obj: Object) => obj[keyGetter] : keyGetter
  let maxKeyCount = 0
  const mutable: UseBindedCallbackMapperState = {
    handler,
    bindArgs,
    mapper: (array: Array<Object>, mapFunction: MapFunction) => {
      ++maxKeyCount
      const mapped = array.map((item, index, array) => {
        const key = getKey(item)
        const itemMutable = itemsMutable[key]
        if (itemMutable) {
          ++itemMutable.count
          itemMutable.item = item
          return mapFunction(key, itemMutable.callback, item, index, array)
        }
        const currentItemMutable: UseCallbackMapperItemState = {
          count: maxKeyCount,
          item,
          callback: function (this: any, ...args: any[]) {
            mutable.handler.call(
              this,
              currentItemMutable.item,
              ...mutable.bindArgs,
              ...args
            )
          }
        }
        itemsMutable[key] = currentItemMutable
        return mapFunction(key, currentItemMutable.callback, item, index, array)
      })
      const deletedKeys = []
      for (const key in itemsMutable) {
        if (itemsMutable[key].count < maxKeyCount) {
          deletedKeys.push(key)
        }
      }
      deletedKeys.forEach((key) => {
        delete itemsMutable[key]
      })
      return mapped
    }
  }
  ref.current = mutable
  return mutable.mapper
}

export default useBindedCallbackMapper
