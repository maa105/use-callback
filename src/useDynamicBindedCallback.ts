import { useRef } from 'react'
import {
  DynamicCallback,
  Handler,
  UseDynamicBindedCallbackState
} from './typings'

const useDynamicBindedCallback = (...bindArgs: any[]) => {
  const ref = useRef<UseDynamicBindedCallbackState>()
  const current = ref.current
  if (current) {
    current.bindArgs = bindArgs
    return current.callback
  }
  const mutable: any = { bindArgs }
  const callback: DynamicCallback = function (this: any, ...args: any[]) {
    return mutable.handler.call(this, ...mutable.bindArgs, ...args)
  }
  callback.setHandler = (handler: Handler) => {
    mutable.handler = handler
    delete callback.setHandler
  }
  mutable.callback = callback
  ref.current = mutable
  return callback
}

export default useDynamicBindedCallback
