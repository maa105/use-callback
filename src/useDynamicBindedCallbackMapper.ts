import { useRef } from 'react'
import {
  DynamicMapper,
  Handler,
  KeyGetter,
  MapFunction,
  UseCallbackMapperItemsState,
  UseCallbackMapperItemState,
  UseDynamicBindedCallbackMapperState
} from './typings'

const useDynamicBindedCallbackMapper = (...bindArgs: any[]): DynamicMapper => {
  const ref = useRef<UseDynamicBindedCallbackMapperState>()
  const current = ref.current
  if (current) {
    current.bindArgs = bindArgs
    return current.mapper
  }
  const itemsMutable: UseCallbackMapperItemsState = {}
  let maxKeyCount = 0
  const mutable: any = {
    bindArgs,
    mapper: (array: Array<Object>, mapFunction: MapFunction) => {
      ++maxKeyCount
      const mapped = array.map((item, index, array) => {
        const key = mutable.getKey(item)
        const itemMutable = itemsMutable[key]
        if (itemMutable) {
          ++itemMutable.count
          itemMutable.item = item
          return mapFunction(key, itemMutable.callback, item, index, array)
        }
        const currentItemMutable: UseCallbackMapperItemState = {
          count: maxKeyCount,
          item,
          callback: function (this: any, ...args: any[]) {
            return mutable.handler?.call(
              this,
              currentItemMutable.item,
              ...mutable.bindArgs,
              ...args
            )
          }
        }
        itemsMutable[key] = currentItemMutable
        return mapFunction(key, currentItemMutable.callback, item, index, array)
      })
      const deletedKeys = []
      for (const key in itemsMutable) {
        if (itemsMutable[key].count < maxKeyCount) {
          deletedKeys.push(key)
        }
      }
      deletedKeys.forEach((key) => {
        delete itemsMutable[key]
      })
      return mapped
    }
  }
  const mapper = mutable.mapper
  mapper.setHandler = (handler: Handler) => {
    mutable.handler = handler
    delete mapper.setHandler
  }
  mapper.setKeyGetter = (keyGetter: string | KeyGetter) => {
    if (typeof keyGetter === 'string') {
      mutable.getKey = (obj: any) => obj[keyGetter]
    } else {
      mutable.getKey = keyGetter
    }
    delete mapper.setKeyGetter
  }
  ref.current = mutable
  return mapper
}

export default useDynamicBindedCallbackMapper
