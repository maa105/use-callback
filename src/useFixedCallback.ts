import { useRef } from 'react'
import { Handler, UseFixedCallbackState } from './typings'

const useFixedCallback = (handler: Handler) => {
  const ref = useRef<UseFixedCallbackState>()
  const current = ref.current
  if (current) {
    current.handler = handler
    return current.callback
  }
  const mutable: any = {
    handler
  }
  const callback = function (this: any, ...args: any[]) {
    return mutable.handler.apply(this, args)
  }
  mutable.callback = callback
  ref.current = mutable
  return callback
}

export default useFixedCallback
