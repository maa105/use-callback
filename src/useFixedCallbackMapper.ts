import { useRef } from 'react'
import {
  Handler,
  KeyGetter,
  MapFunction,
  UseFixedCallbackMapperState,
  UseCallbackMapperItemsState,
  UseCallbackMapperItemState
} from './typings'

const useFixedCallbackMapper = (
  handler: Handler,
  keyGetter: string | KeyGetter
) => {
  const ref = useRef<UseFixedCallbackMapperState>()
  const current = ref.current
  if (current) {
    current.handler = handler
    return current.mapper
  }
  const itemsMutable: UseCallbackMapperItemsState = {}
  const getKey: KeyGetter =
    typeof keyGetter === 'string' ? (obj: Object) => obj[keyGetter] : keyGetter
  let maxKeyCount = 0
  const mutable: UseFixedCallbackMapperState = {
    handler,
    mapper: (array: Array<Object>, mapFunction: MapFunction) => {
      ++maxKeyCount
      const mapped = array.map((item, index, array) => {
        const key = getKey(item)
        const itemMutable = itemsMutable[key]
        if (itemMutable) {
          ++itemMutable.count
          itemMutable.item = item
          return mapFunction(key, itemMutable.callback, item, index, array)
        }
        const currentItemMutable: UseCallbackMapperItemState = {
          count: maxKeyCount,
          item,
          callback: function (this: any, ...args: any[]) {
            mutable.handler.call(this, currentItemMutable.item, ...args)
          }
        }
        itemsMutable[key] = currentItemMutable
        return mapFunction(key, currentItemMutable.callback, item, index, array)
      })
      const deletedKeys = []
      for (const key in itemsMutable) {
        if (itemsMutable[key].count < maxKeyCount) {
          deletedKeys.push(key)
        }
      }
      deletedKeys.forEach((key) => {
        delete itemsMutable[key]
      })
      return mapped
    }
  }
  ref.current = mutable
  return mutable.mapper
}

export default useFixedCallbackMapper
