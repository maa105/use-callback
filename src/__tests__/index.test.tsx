import React, { useRef, useState } from 'react'
import { fireEvent, render } from '@testing-library/react'
import useFixedCallback from '../useFixedCallback'
import useBindedCallback from '../useBindedCallback'
import useDynamicBindedCallback from '../useDynamicBindedCallback'
import useBindedCallbackMapper from '../useBindedCallbackMapper'
import useDynamicBindedCallbackMapper from '../useDynamicBindedCallbackMapper'
import useFixedCallbackMapper from '../useFixedCallbackMapper'

describe('useFixedCallback', () => {
  const initTest = () => {
    const callbacks: any[] = []
    const Child = ({ onClick, children }: any) => {
      return <button onClick={onClick}>{children}</button>
    }
    const TestComponent = ({ onClick }: any) => {
      const [count, setCount] = useState(0)
      const callback = useFixedCallback((e) => {
        e.persist()
        setCount(count + 1)
        return onClick(count, e)
      })
      callbacks.push(callback)
      return (
        <>
          <span data-testid='counter-value'>{count}</span>
          <Child onClick={callback}>Ola</Child>
        </>
      )
    }
    return { callbacks, TestComponent }
  }
  it('callback identiy remains the same', async () => {
    const { callbacks, TestComponent } = initTest()
    const onClick = jest.fn()
    const { findByText } = render(<TestComponent onClick={onClick} />)

    const button = await findByText('Ola')
    fireEvent.click(button)

    expect(callbacks.length).toEqual(2)
    expect(callbacks[0]).toStrictEqual(callbacks[1])
  })

  it('handler is called with correct args', async () => {
    const { TestComponent } = initTest()
    const onClick = jest.fn()
    const { findByText, findByTestId } = render(
      <TestComponent onClick={onClick} />
    )

    const button = await findByText('Ola')
    fireEvent.click(button)

    const span = await findByTestId('counter-value')
    expect(span.innerHTML).toEqual('1')

    expect(onClick.mock.calls[0][0]).toEqual(0)
    expect(onClick.mock.calls[0][1].type).toEqual('click')

    fireEvent.click(button)

    expect(span.innerHTML).toEqual('2')

    expect(onClick.mock.calls[1][0]).toEqual(1)
    expect(onClick.mock.calls[1][1].type).toEqual('click')
  })
})

describe('useBindedCallback', () => {
  const initTest = () => {
    const callbacks: any[] = []
    const Child = ({ onClick, children }: any) => {
      return <button onClick={onClick}>{children}</button>
    }
    const TestComponent = ({ onClick }: any) => {
      const [count, setCount] = useState(0)
      const callback = useBindedCallback(
        (setCount, onClick, count, e) => {
          e.persist()
          setCount(count + 1)
          return onClick(count, e)
        },
        setCount,
        onClick,
        count
      )
      callbacks.push(callback)
      return (
        <>
          <span data-testid='counter-value'>{count}</span>
          <Child onClick={callback}>Ola</Child>
        </>
      )
    }
    return { callbacks, TestComponent }
  }
  it('callback identiy remains the same', async () => {
    const { callbacks, TestComponent } = initTest()
    const onClick = jest.fn()
    const { findByText } = render(<TestComponent onClick={onClick} />)

    const button = await findByText('Ola')
    fireEvent.click(button)

    expect(callbacks.length).toEqual(2)
    expect(callbacks[0]).toStrictEqual(callbacks[1])
  })

  it('handler is called with correct args', async () => {
    const { TestComponent } = initTest()
    const onClick = jest.fn()
    const { findByText, findByTestId } = render(
      <TestComponent onClick={onClick} />
    )

    const button = await findByText('Ola')
    fireEvent.click(button)

    const span = await findByTestId('counter-value')
    expect(span.innerHTML).toEqual('1')

    expect(onClick.mock.calls[0][0]).toEqual(0)
    expect(onClick.mock.calls[0][1].type).toEqual('click')

    fireEvent.click(button)

    expect(span.innerHTML).toEqual('2')

    expect(onClick.mock.calls[1][0]).toEqual(1)
    expect(onClick.mock.calls[1][1].type).toEqual('click')
  })
})

describe('useDynamicBindedCallback', () => {
  const initTest = () => {
    const callbacks: any[] = []
    const Child = ({ onClick, children }: any) => {
      return <button onClick={onClick}>{children}</button>
    }
    const TestComponent = ({ onClick }: any) => {
      const [count, setCount] = useState(0)
      const callback = useDynamicBindedCallback(setCount, onClick, count)
      // eslint-disable-next-line no-unused-expressions
      callback.setHandler?.(
        (
          setCount: Function,
          onClick: Function,
          count: number,
          e: React.MouseEvent
        ) => {
          e.persist()
          setCount(count + 1)
          return onClick(count, e)
        }
      )
      callbacks.push(callback)
      return (
        <>
          <span data-testid='counter-value'>{count}</span>
          <Child onClick={callback}>Ola</Child>
        </>
      )
    }
    return { callbacks, TestComponent }
  }
  it('callback identiy remains the same', async () => {
    const { callbacks, TestComponent } = initTest()
    const onClick = jest.fn()
    const { findByText } = render(<TestComponent onClick={onClick} />)

    const button = await findByText('Ola')
    fireEvent.click(button)

    expect(callbacks.length).toEqual(2)
    expect(callbacks[0]).toStrictEqual(callbacks[1])
  })

  it('handler is called with correct args', async () => {
    const { TestComponent } = initTest()
    const onClick = jest.fn()
    const { findByText, findByTestId } = render(
      <TestComponent onClick={onClick} />
    )

    const button = await findByText('Ola')
    fireEvent.click(button)

    const span = await findByTestId('counter-value')
    expect(span.innerHTML).toEqual('1')

    expect(onClick.mock.calls[0][0]).toEqual(0)
    expect(onClick.mock.calls[0][1].type).toEqual('click')

    fireEvent.click(button)

    expect(span.innerHTML).toEqual('2')

    expect(onClick.mock.calls[1][0]).toEqual(1)
    expect(onClick.mock.calls[1][1].type).toEqual('click')
  })
})

describe('useFixedCallbackMapper', () => {
  const initTest = () => {
    const mappers: any[] = []
    const Child = React.memo(({ onClick, label }: any) => {
      const countRef = useRef(0)
      ++countRef.current
      return (
        <button data-testid='child-button' onClick={onClick}>
          {label}
          <br />
          [RenderCount:{countRef.current}]
        </button>
      )
    })
    const TestComponent = ({ items, onItemClick }: any) => {
      const [count, setCount] = useState(0)
      const mapper = useFixedCallbackMapper(
        (item: Object, e: React.MouseEvent) => {
          e.persist()
          setCount(count + 1)
          return onItemClick(item, count, e)
        },
        'id'
      )
      mappers.push(mapper)
      return (
        <div data-testid='root'>
          <span data-testid='counter-value'>{count}</span>
          {mapper(items, (key: string, callback: Function, item: any) => (
            <Child key={key} onClick={callback} label={item.name} />
          ))}
        </div>
      )
    }
    return { mappers, TestComponent }
  }

  it('callback identiy remains the same', async () => {
    const { mappers, TestComponent } = initTest()
    const onClick = jest.fn()
    const items = [
      { id: 1, name: 'Item1' },
      { id: 2, name: 'Item2' },
      { id: 3, name: 'Item3' }
    ]
    const { findByTestId, findAllByTestId } = render(
      <TestComponent onItemClick={onClick} items={items} />
    )

    const root = await findByTestId('root')
    const buttons = await findAllByTestId('child-button')
    expect(buttons.length).toEqual(3)

    expect(root.outerHTML).toMatchSnapshot()

    fireEvent.click(buttons[0])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }

    fireEvent.click(buttons[1])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }

    fireEvent.click(buttons[2])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }
  })

  it('handler is called with correct args', async () => {
    const { TestComponent } = initTest()
    const onClick = jest.fn()
    const items = [
      { id: 1, name: 'Item1' },
      { id: 2, name: 'Item2' },
      { id: 3, name: 'Item3' }
    ]
    const { findByTestId, findAllByTestId } = render(
      <TestComponent onItemClick={onClick} items={items} />
    )

    const buttons = await findAllByTestId('child-button')
    const span = await findByTestId('counter-value')

    expect(span.innerHTML).toEqual('0')

    fireEvent.click(buttons[0])

    expect(span.innerHTML).toEqual('1')

    expect(onClick.mock.calls[0][0] === items[0]).toEqual(true)
    expect(onClick.mock.calls[0][1]).toEqual(0)
    expect(onClick.mock.calls[0][2].type).toEqual('click')

    fireEvent.click(buttons[2])

    expect(span.innerHTML).toEqual('2')

    expect(onClick.mock.calls[1][0] === items[2]).toEqual(true)
    expect(onClick.mock.calls[1][1]).toEqual(1)
    expect(onClick.mock.calls[1][2].type).toEqual('click')
  })
})

describe('useBindedCallbackMapper', () => {
  const initTest = () => {
    const mappers: any[] = []
    const Child = React.memo(({ onClick, label }: any) => {
      const countRef = useRef(0)
      ++countRef.current
      return (
        <button data-testid='child-button' onClick={onClick}>
          {label}
          <br />
          [RenderCount:{countRef.current}]
        </button>
      )
    })
    const TestComponent = ({ items, onItemClick }: any) => {
      const [count, setCount] = useState(0)
      const mapper = useBindedCallbackMapper(
        (item: Object, count: number, e: React.MouseEvent) => {
          e.persist()
          setCount(count + 1)
          return onItemClick(item, count, e)
        },
        'id',
        count
      )
      mappers.push(mapper)
      return (
        <div data-testid='root'>
          <span data-testid='counter-value'>{count}</span>
          {mapper(items, (key: string, callback: Function, item: any) => (
            <Child key={key} onClick={callback} label={item.name} />
          ))}
        </div>
      )
    }
    return { mappers, TestComponent }
  }

  it('callback identiy remains the same', async () => {
    const { mappers, TestComponent } = initTest()
    const onClick = jest.fn()
    const items = [
      { id: 1, name: 'Item1' },
      { id: 2, name: 'Item2' },
      { id: 3, name: 'Item3' }
    ]
    const { findByTestId, findAllByTestId } = render(
      <TestComponent onItemClick={onClick} items={items} />
    )

    const root = await findByTestId('root')
    const buttons = await findAllByTestId('child-button')
    expect(buttons.length).toEqual(3)

    expect(root.outerHTML).toMatchSnapshot()

    fireEvent.click(buttons[0])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }

    fireEvent.click(buttons[1])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }

    fireEvent.click(buttons[2])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }
  })

  it('handler is called with correct args', async () => {
    const { TestComponent } = initTest()
    const onClick = jest.fn()
    const items = [
      { id: 1, name: 'Item1' },
      { id: 2, name: 'Item2' },
      { id: 3, name: 'Item3' }
    ]
    const { findByTestId, findAllByTestId } = render(
      <TestComponent onItemClick={onClick} items={items} />
    )

    const buttons = await findAllByTestId('child-button')
    const span = await findByTestId('counter-value')

    expect(span.innerHTML).toEqual('0')

    fireEvent.click(buttons[0])

    expect(span.innerHTML).toEqual('1')

    expect(onClick.mock.calls[0][0] === items[0]).toEqual(true)
    expect(onClick.mock.calls[0][1]).toEqual(0)
    expect(onClick.mock.calls[0][2].type).toEqual('click')

    fireEvent.click(buttons[2])

    expect(span.innerHTML).toEqual('2')

    expect(onClick.mock.calls[1][0] === items[2]).toEqual(true)
    expect(onClick.mock.calls[1][1]).toEqual(1)
    expect(onClick.mock.calls[1][2].type).toEqual('click')
  })
})

describe('useDynamicBindedCallbackMapper', () => {
  const initTest = () => {
    const mappers: any[] = []
    const Child = React.memo(({ onClick, label }: any) => {
      const countRef = useRef(0)
      ++countRef.current
      return (
        <button data-testid='child-button' onClick={onClick}>
          {label}
          <br />
          [RenderCount:{countRef.current}]
        </button>
      )
    })
    const TestComponent = ({ items, onItemClick }: any) => {
      const [count, setCount] = useState(0)
      const mapper = useDynamicBindedCallbackMapper(count)
      // eslint-disable-next-line no-unused-expressions
      mapper.setHandler?.(
        (item: Object, count: number, e: React.MouseEvent) => {
          e.persist()
          setCount(count + 1)
          return onItemClick(item, count, e)
        }
      )
      // eslint-disable-next-line no-unused-expressions
      mapper.setKeyGetter?.(({ id }: any) => id)
      mappers.push(mapper)
      return (
        <div data-testid='root'>
          <span data-testid='counter-value'>{count}</span>
          {mapper(items, (key: string, callback: Function, item: any) => (
            <Child key={key} onClick={callback} label={item.name} />
          ))}
        </div>
      )
    }
    return { mappers, TestComponent }
  }

  it('callback identiy remains the same', async () => {
    const { mappers, TestComponent } = initTest()
    const onClick = jest.fn()
    const items = [
      { id: 1, name: 'Item1' },
      { id: 2, name: 'Item2' },
      { id: 3, name: 'Item3' }
    ]
    const { findByTestId, findAllByTestId } = render(
      <TestComponent onItemClick={onClick} items={items} />
    )

    const root = await findByTestId('root')
    const buttons = await findAllByTestId('child-button')
    expect(buttons.length).toEqual(3)

    expect(root.outerHTML).toMatchSnapshot()

    fireEvent.click(buttons[0])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }

    fireEvent.click(buttons[1])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }

    fireEvent.click(buttons[2])

    expect(root.outerHTML).toMatchSnapshot()
    for (let i = 1; i < mappers.length; i++) {
      expect(mappers[0]).toStrictEqual(mappers[i])
    }
  })

  it('handler is called with correct args', async () => {
    const { TestComponent } = initTest()
    const onClick = jest.fn()
    const items = [
      { id: 1, name: 'Item1' },
      { id: 2, name: 'Item2' },
      { id: 3, name: 'Item3' }
    ]
    const { findByTestId, findAllByTestId } = render(
      <TestComponent onItemClick={onClick} items={items} />
    )

    const buttons = await findAllByTestId('child-button')
    const span = await findByTestId('counter-value')

    expect(span.innerHTML).toEqual('0')

    fireEvent.click(buttons[0])

    expect(span.innerHTML).toEqual('1')

    expect(onClick.mock.calls[0][0] === items[0]).toEqual(true)
    expect(onClick.mock.calls[0][1]).toEqual(0)
    expect(onClick.mock.calls[0][2].type).toEqual('click')

    fireEvent.click(buttons[2])

    expect(span.innerHTML).toEqual('2')

    expect(onClick.mock.calls[1][0] === items[2]).toEqual(true)
    expect(onClick.mock.calls[1][1]).toEqual(1)
    expect(onClick.mock.calls[1][2].type).toEqual('click')
  })
})
