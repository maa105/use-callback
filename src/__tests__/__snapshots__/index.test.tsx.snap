// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBindedCallbackMapper callback identiy remains the same 1`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">0</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useBindedCallbackMapper callback identiy remains the same 2`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">1</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useBindedCallbackMapper callback identiy remains the same 3`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">2</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useBindedCallbackMapper callback identiy remains the same 4`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">3</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useDynamicBindedCallbackMapper callback identiy remains the same 1`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">0</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useDynamicBindedCallbackMapper callback identiy remains the same 2`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">1</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useDynamicBindedCallbackMapper callback identiy remains the same 3`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">2</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useDynamicBindedCallbackMapper callback identiy remains the same 4`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">3</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useFixedCallbackMapper callback identiy remains the same 1`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">0</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useFixedCallbackMapper callback identiy remains the same 2`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">1</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useFixedCallbackMapper callback identiy remains the same 3`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">2</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;

exports[`useFixedCallbackMapper callback identiy remains the same 4`] = `"<div data-testid=\\"root\\"><span data-testid=\\"counter-value\\">3</span><button data-testid=\\"child-button\\">Item1<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item2<br>[RenderCount:1]</button><button data-testid=\\"child-button\\">Item3<br>[RenderCount:1]</button></div>"`;
